### ISSUE TITLE
Can I implement Android notification channels using the Unity SDK?

### ISSUE THREAD

**Issue by JimmyDeemo** [Pushwoosh/pushwoosh-unity#44]
Is it possible to implement the channel setup using C# and the Unity SDK? I am looking for the functionality detailed here; https://www.pushwoosh.com/docs/notification-channels.

If not, will it be implemented or do I need to write a custom native plugin?

**Comment by wfhm** [comment]
@JimmyDeemo hey,

Right now it is only possible with a custom NotificationFactory, which can only be created in a custom native plugin. We'll need to check if it can be implemented for Unity plugin.

Stay tuned!

**Comment by wfhm** [comment]
@JimmyDeemo 

Could you please describe your use-case in a bit more details? Right now you can specify a Channel as per the guide below manually when sending a notification:

https://www.pushwoosh.com/docs/notification-channels

Android OS "remembers" a priority and sound/vibration preferences when a push assigned to a new Channel is received, and, when receiving notifications to this Channel next time, will use these prefs regardless of priority, sound and vibration settings you specify with a next push.

Basically you should just manage already created Channels on your end, and, in this case, there should be no code changes made on application (or Pushwoosh plugin) side.

**Comment by JimmyDeemo** [comment]
One of the uses of Unity is to avoid having to delve into the native side of mobile development, although it is frequently required for more custom behaviour.

As I understand it, when utilizing the Pushwoosh SDK, there is no way to use channels without creating the channels in native code as laid out here; https://developer.android.com/training/notify-user/channels#java.

I wondered if there was facility to create these channels without having to write native code. Perhaps some sort of configuration or initialization calls through Pushwoosh. If this is out of scope for Pushwoosh then feel free to close this issue.

**Comment by wfhm** [comment]
@JimmyDeemo 

Actually if you send a push with 
`"android_root_params":{"pw_channel": "CHANNEL NAME"}`, a notification channel with a specified name will automatically be created (or, if it already exists, a push will be assigned to this channel). I've just tested it and it works with Pushwoosh Unity plugin just fine.

If you only need to create channels and send pushes to them, this functionality should be enough, however if you need more control, please let us know more about the use-case.

**Comment by JimmyDeemo** [comment]
Thanks for the info @wfhm.

This should serve to place our Pushes into channels fine, which is great, but one immediate problem I see is localization. We support over 10 languages for our game and I believe that setting up the channels natively is the area that you provide the localized names for the channels.

I suppose this might be possible to support via the API (without SDK modification) but what about via the Dashboard? Can the root params property be multiplexed across different languages in the presets?

**Comment by wfhm** [comment]
@JimmyDeemo I think we can release a microplugin to add notification channels for Unity. Will discuss it with the team today, meanwhile could you please describe which functionality, apart from localizing Channel names you want to achieve? Ideally, having your usecase would help us a lot.

**Comment by JimmyDeemo** [comment]
For our use case all we want is to be able to define a channel that a notification should be in, so that users can choose which notifications they want on/off in the OS's settings screen. This channel name would need to be localized such that it appears in the same language the push notification is sent in.

I know that some apps like to use channels for setting an importance level, which would be useful. e.g. Key pushes for the project would vibrate the phone, make noise, lower priority pushes would drop vibrate.

That's really all we are looking for.

**Comment by wfhm** [comment]
@JimmyDeemo Thanks for clarification! Will come back to you with a news shortly.

**Comment by wfhm** [comment]
@JimmyDeemo This functionality is finally added with the latest release.

To customize your channel name and its description, you should do the following:

1. Send a push with `"android_root_params":{"pw_channel":"defaultchannelname"}
2. Create a class `MyNotificationChannelDelegate : NotificationChannelDelegate`
3. Override its `ChannelName` and `ChannelDescription` classes - these method receive channel name sent as "pw_channel" parameter, so you can implement your own logics to modify it, e.g. by checking device language and setting a localized version.

Public documentation will be updated shortly.
